cmd/cosign/cli/verify/verify_bundle.go

```diff
@@ -0,0 +1,163 @@
+	bundle, err := sgbundle.LoadJSONFromPath(bundlePath)
+	if err != nil {
+		return err
+	}
+
+	var trustedroot *root.TrustedRoot
+
+	if trustedRootPath == "" {
+		// Assume we're using public good instance; fetch via TUF
+		trustedroot, err = root.FetchTrustedRoot()
+		if err != nil {
+			return err
+		}
+	} else {
+		trustedroot, err = root.NewTrustedRootFromPath(trustedRootPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	trustedmaterial := &verifyTrustedMaterial{TrustedMaterial: trustedroot}
+
+	// See if we need to wrap trusted root with provided key
+	if keyRef != "" {
+		signatureVerifier, err := sigs.PublicKeyFromKeyRef(ctx, keyRef)
+		if err != nil {
+			return err
+		}
+
+		newExpiringKey := root.NewExpiringKey(signatureVerifier, time.Time{}, time.Time{})
+		trustedmaterial.keyTrustedMaterial = root.NewTrustedPublicKeyMaterial(func(_ string) (root.TimeConstrainedVerifier, error) {
+			return newExpiringKey, nil
+		})
+	} else if sk {
+		s, err := pivkey.GetKeyWithSlot(slot)
+		if err != nil {
+			return fmt.Errorf("opening piv token: %w", err)
+		}
+		defer s.Close()
+		signatureVerifier, err := s.Verifier()
+		if err != nil {
+			return fmt.Errorf("loading public key from token: %w", err)
+		}
+
+		newExpiringKey := root.NewExpiringKey(signatureVerifier, time.Time{}, time.Time{})
+		trustedmaterial.keyTrustedMaterial = root.NewTrustedPublicKeyMaterial(func(_ string) (root.TimeConstrainedVerifier, error) {
+			return newExpiringKey, nil
+		})
+	}
+
+	identityPolicies := []verify.PolicyOption{}
+
+	verificationMaterial := bundle.GetVerificationMaterial()
+
+	if verificationMaterial == nil {
+		return fmt.Errorf("no verification material in bundle")
+	}
+
+	if verificationMaterial.GetPublicKey() != nil {
+		identityPolicies = append(identityPolicies, verify.WithKey())
+	} else {
+		sanMatcher, err := verify.NewSANMatcher(certIdentity, certIdentityRegexp)
+		if err != nil {
+			return err
+		}
+
+		issuerMatcher, err := verify.NewIssuerMatcher(certOIDCIssuer, certOIDCIssuerRegex)
+		if err != nil {
+			return err
+		}
+
+		extensions := certificate.Extensions{
+			GithubWorkflowTrigger:    githubWorkflowTrigger,
+			GithubWorkflowSHA:        githubWorkflowSHA,
+			GithubWorkflowName:       githubWorkflowName,
+			GithubWorkflowRepository: githubWorkflowRepository,
+			GithubWorkflowRef:        githubWorkflowRef,
+		}
+
+		certIdentity, err := verify.NewCertificateIdentity(sanMatcher, issuerMatcher, extensions)
+		if err != nil {
+			return err
+		}
+
+		identityPolicies = append(identityPolicies, verify.WithCertificateIdentity(certIdentity))
+	}
+
+	// Make some educated guesses about verification policy
+	verifierConfig := []verify.VerifierOption{}
+
+	if len(trustedroot.RekorLogs()) > 0 && !ignoreTlog {
+		verifierConfig = append(verifierConfig, verify.WithTransparencyLog(1), verify.WithObserverTimestamps(1))
```

Comment:
Should we use `WithIntegratedTimestamps` instead of `WithObserverTimestamps`?

Reply:
This is a good idea - in other contexts when I'm using sigstore-go I'll use `WithObserverTimestamps` to take advantage of a signed timestamp if it's available, but I think you're right that in cosign if you aren't using `--insecure-ignore-tlog`, you probably do want to specifically check the integrated timestamp.